grafana/grafana
# Datasources: Improve error handling for error messages

**What this PR does / why we need it**:

This PR improves the way we handle errors when testing a data source results in failure (when clicking on "Test & Save" button).

So far the logic was to return `error.statusText` and fall back to `error.message` if no `statusText` is defined. In practice `error.statusText` is a more generic message and `error.message` should provide more detailed information. What is more the error might be returned in the response body in `error.data.message`. This PR changes that the error message is extracted from the first non-empty property in the following order:

- `error.message`
- `error.data.message`
- `error.statusText`

This should give more precise information about what happened.



**Which issue(s) this PR fixes**:



Fixes #34013

Backend changed the logic so in case of authentication errors  HTTP 401 is no longer returned but more appropriate in this case HTTP 400 with a message in the response body (see the issue for the explanation). This, however, means that the logic described above is picking up the status code for 400 (Bad Request) without additional info. By changing the way the error message is extracted we will get a more meaningful error.

**How to test it?**

Run `make devenv sources=graphite`. Create a Graphite data source. For given URLs you should get the following errors:

| URL | Error message |
|-----|----------------|
| https://graphite-us-central1.grafana.net/graphite | Authentication to data source failed |
| http://bad-gateway | Bad Gateway | 
| http://localhost:8180 | Data source is working |

for example:

<img width="660" alt="Screenshot 2021-06-02 at 17 21 54" src="https://user-images.githubusercontent.com/745532/120507423-16998480-c3c7-11eb-911c-b24dc6d49b2f.png">


**Special notes for your reviewer**:

N/A
